File: LeetCode/1419. Minimum Number of Frogs Croaking/solution.py
# class Solution:
#     def minNumberOfFrogs(self, croakOfFrogs: str) -> int:
#         croak = "croak"
#         res = 0
#         s = list(map(lambda x: croak.index(x), croakOfFrogs))
#         cc = [0]*5
#         for x in s:
#             cc[x] += 1
#             if x > 0:
#                 cc[x-1] -= 1
#                 if cc[x-1] < 0:
#                     return -1
#             print(cc)
#             res = max(res,sum(cc[:4]))
#         if any(cc[i] != 0 for i in range(4)):
#             return -1
#         return res

class Solution:
    def minNumberOfFrogs(self, croakOfFrogs: str) -> int:
        res = 0
        state = [0]*5
        for ch in croakOfFrogs:
            cur = "croak".index(ch)
            pre = (cur-1)%5
            state[cur] += 1
            state[pre] -= 1
            if state[pre] < 0:
                if pre == 4:
                    res += 1
                    state[pre] += 1
                else:
                    return -1
        return res if state[-1] == res else -1

File: LeetCode/1419. Minimum Number of Frogs Croaking/solution.cpp
    class Solution {
    public:
        int minNumberOfFrogs(string croakOfFrogs) {
            vector<int> mp(128), cnt(5,0);
            int res = 0;
            string str = "croak";
            for (int i = 0; i < str.length(); i++)
                mp[str[i]] = i+1;
            for (auto ch : croakOfFrogs){
                int s = mp[ch];
                cnt[s%5] += 1;
                cnt[s-1] -= 1;
                if (s-1 > 0 && cnt[s-1] < 0)
                    return -1;
                res = max(res, -cnt[0]);
            }
            return cnt[0] == 0 ? res : -1;
        }
    };